### ISSUE TITLE
IndexError: index -1 is out of bounds for axis 0 with size 0

### ISSUE THREAD

**Issue by charlie9526** [SeldonIO/alibi#317]
```
~/.local/lib/python3.6/site-packages/alibi/explainers/anchor_tabular.py in perturbation(self, anchor, num_samples)
    278         )
    279         nb_partial_anchors = np.array([len(n_records) for n_records in reversed(partial_anchor_rows)])
--> 280         coverage = nb_partial_anchors[-1] / self.n_records
    281 
    282         # if there are enough train records containing the anchor, replace the original records and return...

IndexError: index -1 is out of bounds for axis 0 with size 0
```

**Comment by jklaise** [comment]
Looks like the the variable `nb_partial_anchors` is empty, this could theoretically happen if there are no rows in the training set where the proposed anchor holds. This could be a bug on our side as we should catch that error if that is what is happening.

Can you run the code in a debugger and print out the variable values? Probably starting with `allowed_rows` in this method: https://github.com/SeldonIO/alibi/blob/ce4d695285236fe1b638446bedcabed5e8122d1f/alibi/explainers/anchor_tabular.py#L237-L306

**Comment by charlie9526** [comment]
Thank you for the quick reply. In the current situation, I am not using a debugger. I will upload them later.
Thank you!

**Comment by enricorotundo** [comment]
Fixed in `alibi>=0.5.6` ?

https://github.com/SeldonIO/alibi/blob/9af1d73d046a2239ce73c12c2e1f835fecc89780/alibi/explainers/anchor_tabular.py#L281

**Comment by jklaise** [comment]
@enricorotundo I don't think so because you would get the same error regardless if the index was 0 or -1.

I think the issue may originate because the list is empty (meaning `partial_anchor_rows` is empty). Is this something you've also experienced?

**Comment by enricorotundo** [comment]
Yes but I've workaround it by tweaking my dataset a bit so I get different anchors and it wouldn't fail.

**Comment by jklaise** [comment]
@enricorotundo can you elaborate a little bit? Maybe we can get to the bottom of this and implement appropriate handling.

**Comment by enricorotundo** [comment]
@jklaise I'm working with a customer dataset using AnchorTabular via Seldon so I'm not so sure how to provide a reproducible example. Anyhow, replacing `AnchorTabular(clf.predict(x))` with `AnchorTabular(clf.predict_proba(x))` magically works.

**Comment by jklaise** [comment]
@enricorotundo hmm that's interesting, we do internal processing to convert models outputing probabilities to ones outputing labels using an `ArgmaxTransformer`, I wonder if there is some bug related to that: https://github.com/SeldonIO/alibi/blob/5f1be275f8d57fac83cee0e97bc5dbaf7a95e501/alibi/explainers/anchor_tabular.py#L961-L968

**Comment by jklaise** [comment]
Needs reproducing to proceed.

**Comment by jklaise** [comment]
I have now reproduced this and one way this happens is when the instance to be explained has a categorical value that is not represented in the training data.

I will post a minimal example and some possible solutions. The easy option would be checking whether all categorical values (and numerical bins?) of the instance are present in the training set and raising an Error / returning an "empty" explanation if so. Perhaps there is a workaround so we can still sample, but would need to think more about it.
